argus/environment.py

```diff
@@ -122,21 +148,55 @@ class DevstackEnvironmentPreparer(BaseOpenstackEnvironmentPreparer):
     # since the parent methods are actual methods.
     # pylint: disable=arguments-differ
 
+    def _wait_for_nova_services(self):
+        # Wait until the nova services are up again
+        LOG.info("Waiting for the services to be up again...")
+
+        cmd = 'openstack compute service list -f csv -c State --quote none'
+        while True:
+            stdout = self._run_command(cmd)
+            statuses = stdout[1:]
+            if all(entry == "up"
+                   for entry in statuses):
+                break
+
     @staticmethod
-    def _wait_for_nova_services():
+    def _wait_for_api():
+        LOG.info("Waiting for the API to be up...")
+
+        username = os.environ['OS_USERNAME']
+        password = os.environ['OS_PASSWORD']
+        auth = os.environ['OS_AUTH_URL']
+        tenant = os.environ['OS_TENANT_NAME']
+
+        client = nova.Client(username, password, tenant, auth)
+        while True:
+            try:
+                client.images.list()
+                break
+            except Exception:
+                time.sleep(1)
+
+
+class RDOEnvironmentPreparer(BaseOpenstackEnvironmentPreparer):
+    """An environment preparer for RDO hosts.
+
+    This preparer knows how to patch a configuration file
+    and how to start and stop the openstack services which are
+    running on the host.
+    """
+    # The following are staticmethods, disable this warning
+    # since the parent methods are actual methods.
+    # pylint: disable=arguments-differ
+
+    def _wait_for_nova_services(self):
         # Wait until the nova services are up again

```

Comment:
You're actually duplicating wait_for_nova_services, since it can be found in DevstackEnvironmentPreparer as well. I think that in this case we can move the implementation in the base class.